flink-table/flink-table-common/src/main/java/org/apache/flink/table/connector/source/abilities/SupportsReadingMetadata.java

```diff
@@ -0,0 +1,127 @@
+ * their corresponding data types that the source exposes to the planner. The planner will use this
+ * information for validation and insertion of explicit casts if necessary.
+ *
+ * <p>The planner will select required metadata columns (i.e. perform projection push down) and will
+ * call {@link #applyReadableMetadata(List, DataType)} with a list of metadata keys. An implementation
+ * must ensure that metadata columns are appended at the end of the physical row in the order of the
+ * provided list after the apply method has been called.
+ *
+ * <p>Note: The final output data type emitted by a source changes from the physically produced data
+ * type to a data type with metadata columns. {@link #applyReadableMetadata(List, DataType)} will pass
+ * the updated data type for convenience. If a source implements {@link SupportsProjectionPushDown},
+ * the projection must be applied to the physical data in the first step. The passed updated data type
+ * will have considered information from {@link SupportsProjectionPushDown} already.
+ *
+ * <p>The metadata column's data type must match with {@link #listReadableMetadata()}. For the examples
+ * above, this means that a table source for `t2` returns a TIMESTAMP and not INT. The casting to INT
+ * will be performed by the planner in a subsequent operation:
+ *
+ * <pre>{@code
+ *   // for t1 and t2
+ *   ROW < i INT, s STRING, d DOUBLE >                                              // physical output
+ *   ROW < i INT, s STRING, d DOUBLE, timestamp TIMESTAMP(3) WITH LOCAL TIME ZONE > // final output
+ * }</pre>
+ */
+@PublicEvolving
+public interface SupportsReadingMetadata {
+
+	/**
+	 * Returns the map of metadata keys and their corresponding data types that can be produced by this
+	 * table source for reading.
+	 *
+	 * <p>The returned map will be used by the planner for validation and insertion of explicit casts
+	 * (see {@link LogicalTypeCasts#supportsExplicitCast(LogicalType, LogicalType)}) if necessary.
+	 *
+	 * <p>The iteration order of the returned map determines the order of metadata keys in the list
+	 * passed in {@link #applyReadableMetadata(List, DataType)}. Therefore, it might be beneficial to
+	 * return a {@link LinkedHashMap} if a strict metadata column order is required.
+	 *
+	 * <p>If a source forwards metadata from one or more formats, we recommend the following column
+	 * order for consistency:
+	 *
+	 * <pre>{@code
+	 *   KEY FORMAT METADATA COLUMNS + VALUE FORMAT METADATA COLUMNS + SOURCE METADATA COLUMNS
+	 * }</pre>
+	 *
+	 * <p>Metadata key names follow the same pattern as mentioned in {@link Factory}. In case of duplicate
+	 * names in format and source keys, format keys shall have higher precedence.
+	 *
+	 * <p>Regardless of the returned {@link DataType}s, a metadata column is always represented using
+	 * internal data structures (see {@link RowData}).
+	 *
+	 * @see DecodingFormat#listReadableMetadata()
+	 */
+	Map<String, DataType> listReadableMetadata();
```

Comment:
Returning a `Map` might be a bit dangerous when the order is important. A `List` of tuples would seem better. But I'm sure this was discussed in the FLIP. So please ignore... 😅

Reply:
The order of the `Map` is not super important. In the end, the passed `List` decides about the order. But an implementation can sync both methods and avoid index mapping if desired.